src/file_pipe_log/pipe.rs

```diff
@@ -272,31 +371,73 @@ impl<F: FileSystem> SinglePipe<F> {
+        files.fds.len() * self.target_file_size
     }
 
     fn rotate(&self) -> Result<()> {
         self.rotate_imp(&mut self.active_file.lock())
     }
 
     fn purge_to(&self, file_seq: FileSeq) -> Result<usize> {
-        let (purged, remained) = {
+        let (
+            first_purge_seq, /* first seq for purging */
+            purged,          /* count of purged files */
+            recycled,        /* count of recycled files */
+            remained,        /* count of remained files */
+        ) = {
             let mut files = self.files.write();
-            if file_seq > files.active_seq {
+            if file_seq >= files.first_seq + files.fds.len() as u64 {
                 return Err(box_err!("Purge active or newer files"));
             }
-            let end_offset = file_seq.saturating_sub(files.first_seq) as usize;
-            files.fds.drain(..end_offset);
-            files.first_seq = file_seq;
-            (end_offset, files.fds.len())
+            let mut first_purge_seq = 0;
+            let end_offset: usize;
+            let mut recycled: usize = 0;
+            if files.capacity == 0 {
+                // Not capable for Recycle
+                first_purge_seq = files.first_seq;
+                end_offset = file_seq.saturating_sub(files.first_seq) as usize;
+                files.fds.drain(..end_offset);
+                files.first_seq = file_seq;
+                files.first_seq_in_use = file_seq;
+            } else {
+                // Capable for recycling log files. It means that the
+                // FileCollections has a finite volume for storing recycled files.
+                // [1] if `reset_volume` <= `expected_purge_count`, it means
+                //     that the FileCollection for recycling can not recycle
+                //     all files which need to be purged, we should purge
+                //     several files and recycle the others.
+                // [2] if `reset_volume` > `expected_purge_count`, it means
+                //     that the FileCollection has enough space for recycling
+                //     all files which need to be purged.
+                let reset_volume = (files.capacity as u64)
+                    .saturating_sub(files.first_seq_in_use - files.first_seq)
+                    as usize;
+                let expected_purge_count = file_seq.saturating_sub(files.first_seq_in_use) as usize;
+                if expected_purge_count >= reset_volume {
+                    first_purge_seq = files.first_seq;
+                    end_offset = expected_purge_count - reset_volume;
+                    recycled = reset_volume;
+                    files.first_seq += end_offset as u64;
+                    files.first_seq_in_use = file_seq;
+                } else {
+                    end_offset = 0;
+                    recycled = file_seq.saturating_sub(files.first_seq_in_use) as usize;
+                    files.first_seq_in_use = file_seq;
+                }
```

Comment:
Hard to understand, how about:
```
files.first_seq_in_use = file_seq;
purged = std::cmp::min(files.capacity - files.fds.len(), file.first_seq_in_use - files.first_seq);
files.fds.drain(..purged);
```

Reply:
Referring to your advice, the redundant logic has been simplified.